### PR TITLE
Excise Cassette

### PR DESCRIPTION
**Pull request by vchuravy**

### CODE REVIEW

File: lib/CUDAKernels/src/CUDAKernels.jl [hunk 1]
```diff
@@ -335,15 +293,15 @@ end
 # - private memory for each workitem
 ###
 
-@inline function Cassette.overdub(::CUDACtx, ::typeof(Scratchpad), ctx, ::Type{T}, ::Val{Dims}) where {T, Dims}
+@device_override @inline function Scratchpad(ctx, ::Type{T}, ::Val{Dims}) where {T, Dims}
     MArray{__size(Dims), T}(undef)
 end
 
 @inline function Cassette.overdub(::CUDACtx, ::typeof(__synchronize))
```
Comment (leios) [inline lib/CUDAKernels/src/CUDAKernels.jl:300]:
```suggestion
@device_override @inline function __synchronize()
```

Seems to work in my tests

**Comment by leios** [comment]
I just tested and this fixes #286,

**Comment by vchuravy** [comment]
Ok CI seems to be since we need, https://github.com/JuliaGPU/CUDA.jl/pull/1361

File: lib/ROCKernels/src/ROCKernels.jl [hunk 2]
```diff
@@ -206,54 +204,51 @@ function (obj::Kernel{ROCDevice})(args...; ndrange=nothing, dependencies=nothing
     # Launch kernel
     event = AMDGPU.@roc(groupsize=threads, gridsize=nblocks*threads, queue=queue,
                         name=String(nameof(obj.f)), # TODO: maxthreads=maxthreads,
-                        Cassette.overdub(ROCCTX, obj.f, ctx, args...))
+                        obj.f, ctx, args...)
```
Comment (jpsamaroo) [inline lib/ROCKernels/src/ROCKernels.jl:207]:
```suggestion
                        obj.f(ctx, args...))
```

**Comment by vchuravy** [comment]
So at this point the only failure besides KernelGradient is `gamma`!

**Comment by jpsamaroo** [comment]
Final failure should be fixed by: https://github.com/JuliaGPU/AMDGPU.jl/pull/197

File: lib/ROCKernels/Project.toml [hunk 3]
```diff
@@ -12,10 +12,8 @@ SpecialFunctions = "276daf66-3868-5448-9aa4-cd146d93841b"
 StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
 
 [compat]
-AMDGPU = "0.2.5"
-Adapt = "0.4, 1.0, 2.0, 3.0"
-Cassette = "0.3.3"
-KernelAbstractions = "0.7"
-SpecialFunctions = "0.10, 1.0, 2.0"
+AMDGPU = "0.3"
```
Comment (jpsamaroo) [inline lib/ROCKernels/Project.toml:15]:
```suggestion
AMDGPU = "0.3.2"
```

File: lib/CUDAKernels/Project.toml [hunk 4]
```diff
@@ -1,21 +1,17 @@
...
-SpecialFunctions = "276daf66-3868-5448-9aa4-cd146d93841b"
 StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
 
 [compat]
 Adapt = "3.0"
-CUDA = "3.5"
-Cassette = "0.3.3"
-KernelAbstractions = "0.7"
-SpecialFunctions = "0.10, 1.0, 2.0"
+CUDA = "3.5, 3.6, 3.7, 3.8"
```
Comment (jpsamaroo) [inline lib/CUDAKernels/Project.toml:14]:
```suggestion
CUDA = "3.8.2"
```

**Comment by jpsamaroo** [comment]
The reflection failures are the same as for CUDAKernels.

**Pull request merged**